staging/src/k8s.io/apiserver/pkg/audit/context.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package audit
+
+import (
+	"context"
+
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// The key type is unexported to prevent collisions
+type key int
+
+const (
+	// auditAnnotationsKey is the context key for the audit annotations.
+	auditAnnotationsKey key = iota
+)
+
+// annotations = *[]annotation instead of a map to preserve order of insertions
+type annotation struct {
+	key, value string
+}
+
+// WithAuditAnnotations returns a new context that can store audit annotations
+// via the AddAuditAnnotation function.  This function is meant to be called from
+// an early request handler to allow all later layers to set audit annotations.
+// This is required to support flows where handlers that come before WithAudit
+// (such as WithAuthentication) wish to set audit annotations.
+func WithAuditAnnotations(parent context.Context) context.Context {
```

Comment:
How come this can't just be called from `AddAuditAnnotation` when required? Is it because the adding a value to the context requires creating a new context?

Reply:
We cannot modify the context or request in place so we have to wrap at an earlier layer.

---

https://golang.org/pkg/net/http/#Request.WithContext

```go
func (r *Request) WithContext(ctx context.Context) *Request
```

```
WithContext returns a shallow copy of r with its context changed to ctx. The provided ctx must be non-nil.

...
```

---

https://pkg.go.dev/context#WithValue

```go
func WithValue(parent Context, key, val interface{}) Context
```

```
WithValue returns a copy of parent in which the value associated with key is val.

...
```